### ISSUE TITLE
StatefulSet PVC are not zone aware

### ISSUE THREAD

**Issue by linki** [instaclustr/cassandra-operator#304]
I tried out the zone-aware configuration on AWS and it failed because volumes were created in the wrong zone.

Kubernetes version: `v1.14.6`
cassandra-operator: `v2.0.0`

The non-zone aware configuration worked fine. It put 2 out of 3 Pods in the same zone but it worked.

The zone-aware setup looks more robust so I tried the following config:
```yaml
  nodes: 3
  racks:
  - name: eu-central-1a
    labels:
      failure-domain.beta.kubernetes.io/zone: eu-central-1a
  - name: eu-central-1b
    labels:
      failure-domain.beta.kubernetes.io/zone: eu-central-1b
  - name: eu-central-1c
    labels:
      failure-domain.beta.kubernetes.io/zone: eu-central-1c
```

The node in zone `1a` spawned fine but nodes in `1b` and `1c` couldn't be scheduled due to zone volume conflicts. The PV for the `1b` PVC landed in zone `1a`.

I also tried using `nodeAffinity` instead of `nodeSelector` on the cassandra `StatefulSet` but it didn't help.

After talking to a collegue this seems to be a shortcoming of the `StatefulSet` controller itself and not of `cassandra-operator`.

One way to fix it here would be to pre-create three individual `StorageClass`es for each zone. We could then add a key to the configuration above defining which storage class to use for each rack configuration, e.g.
```yaml
  nodes: 3
  racks:
  - name: eu-central-1a
    labels:
      failure-domain.beta.kubernetes.io/zone: eu-central-1a
      storage-class: default-1a
    ...
```

What are your thoughts about this?

Your examples use zone configurations for Google Cloud. Have you seen theses problems in your testing with the zone aware configuration?

**Comment by benbromhead** [comment]
@smiklosovic and @alourie I wonder if we are giving enough flexibility in volume creation here as hinted at by @linki ? iirc we just have a single storage class definition for the entire DC

**Comment by alourie** [comment]
@benbromhead @linki 

TL;DR: set up a storage class to use for the Cassandra pods (for example, `topology-aware-standard`), use `WaitForFirstCustomer` volumeBindingMode and then use this storage class as a parameter to the CRD:

```yaml
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: topology-aware-standard
provisioner: kubernetes.io/aws-ebs
volumeBindingMode: WaitForFirstConsumer
parameters:
  type: gp2
  fsType: ext4 
```
```yaml
...
dataVolumeClaimSpec:
    accessModes:
      - ReadWriteOnce
    storageClassName: topology-aware-standard
    resources:
      requests:
        storage: 500Mi
...
```

Longer explanation: The dynamic volume provisioning can be solved by using `WaitForFirstCustomer` [volumeBindingMode](https://kubernetes.io/docs/concepts/storage/storage-classes/#volume-binding-mode) in a specific storage class used for the Cassandra pods . Basically, we do have it described in the docs, but, at the moment, it's specific to GKE. I'll update the Operations guide to mention that this option is relevant for many types of dynamic volume provisioning (including `AmazonEBS` `GCEPersistentDisk` and `AzureDisk`).

Even more information is at [EBS storage classes](https://docs.aws.amazon.com/eks/latest/userguide/storage-classes.html) and general [Kubernetes storage-class information](https://kubernetes.io/docs/concepts/storage/storage-classes/#parameters).

**Comment by linki** [comment]
@benbromhead @alourie Thanks for the quick responses. I'll try it out and come back to you.

**Comment by alourie** [comment]
@linki Thanks! If it works as expected, then I hope you find the docs update in #305 helpful for the future reference.

**Comment by linki** [comment]
@alourie Using `volumeBindingMode: WaitForFirstConsumer` fixes this issue and StatefulSet's pods are correctly scheduled with their volumes. I wasn't aware of that setting.

However, `WaitForFirstConsumer` doesn't work well with [cluster-autoscaling](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) as new Cassandra Pods don't correctly trigger scale-up in our setup.

But that's a different issue. Closing, thanks for your help.
